src/java/org/apache/cassandra/db/marshal/StringType.java

```diff
@@ -18,10 +18,29 @@
 
 package org.apache.cassandra.db.marshal;
 
+import org.apache.cassandra.utils.ByteBufferUtil;
+
+import java.io.UnsupportedEncodingException;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+
 public abstract class StringType extends AbstractType<String>
 {
     protected StringType(ComparisonType comparisonType)
     {
         super(comparisonType);
     }
+
+    public abstract String charsetName();
+
+    public ByteBuffer concat(StringType leftType,
+                             ByteBuffer left,
+                             StringType rightType,
+                             ByteBuffer right) throws UnsupportedEncodingException
+    {
+        String leftS = new String(left.array(), leftType.charsetName());
+        String rightS = new String(right.array(), rightType.charsetName());
+
+        return ByteBufferUtil.bytes(leftS + rightS);
+    }
```

Comment:
{{ByteBuffer}} do not necessarily have array. If they direct buffer that code will not work. To get the `String`  value from the `ByteBuffer` you should use: `leftType.compose(left)` to convert a `String` back into a `ByteBuffer` you should use the `decompose` method.   ` return decompose(leftS + rightS);`